### PR TITLE
center layout

### PR DESCRIPTION
**Pull request by willmcgugan**
- Implements a "center" layout.
- Fixes an issues with scrollbars. If the mouse was captured when a scrollbar was hidden due to resizing the window, it was impossible to use the other scrollbar.
- Fixes an issue with refresh(layout=True)

https://user-images.githubusercontent.com/554369/183286651-631fa4e4-e2a2-4a8a-9def-420adc8958b3.mov

### CODE REVIEW

File: src/textual/layout.py [hunk 1]
```diff
@@ -1,23 +1,37 @@
 from .widget import Widget
 
 
-class Vertical(Widget):
-    """A container widget to align children vertically."""
+class Container(Widget):
```
Comment (willmcgugan) [inline src/textual/layout.py:4]: A base class for common styles, and for containers where you might want to dynamically tweak the layout.

File: src/textual/layouts/factory.py [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
+
 from .._layout import Layout
-from ..layouts.vertical import VerticalLayout
+from .horizontal import HorizontalLayout
+from .vertical import VerticalLayout
+from .center import CenterLayout
 
 
-LAYOUT_MAP = {
+LAYOUT_MAP: dict[str, type[Layout]] = {
```
Comment (willmcgugan) [inline src/textual/layouts/factory.py:9]: Fixes a typing error below. Mypy couldn't figure this one out by itself.

File: src/textual/scrollbar.py [hunk 3]
```diff
@@ -231,10 +231,14 @@ def render(self) -> RenderableType:
             style=scrollbar_style,
         )
 
-    async def on_enter(self, event: events.Enter) -> None:
+    def on_hide(self, event: events.Hide) -> None:
```
Comment (willmcgugan) [inline src/textual/scrollbar.py:234]: If the scrollbar is hidden it releases the mouse. I'm wondering if this should be standard functionality, but there are some exotic use cases for a hidden widget capturing the mouse.

File: src/textual/widget.py [hunk 4]
```diff
@@ -1040,7 +1040,9 @@ def refresh(
 
         if layout:
             self._layout_required = True
-            self._clear_arrangement_cache()
+            if isinstance(self._parent, Widget):
```
Comment (willmcgugan) [inline src/textual/widget.py:1043]: When a widget's dimensions / visibility change, we want to clear the *parent's* arrangement cache.

**Review by darrenburns — APPROVED** [review]

**Pull request merged**